erts/emulator/beam/dist.c

```diff
@@ -5887,50 +5946,146 @@ BIF_RETTYPE nodes_1(BIF_ALIST_1)
+            eni++;
+        }
+    }
 
+    if (not_connected) {
+        for (dep = erts_not_connected_dist_entries; dep; dep = dep->next) {
+            if (dep != erts_this_dist_entry) {
+                eni->name = dep->sysname;
+                eni->type = am_known;
+                eni->cid =  ~((Uint32) 0);
+                eni++;
+            }
+        }
+        for (dep = erts_pending_dist_entries; dep; dep = dep->next) {
+            eni->name = dep->sysname;
+            eni->type = am_known;
+            eni->cid =  ~((Uint32) 0);
+            eni++;
+        }
+    }
+
+    erts_rwmtx_runlock(&erts_dist_table_rwmtx);
+
+    eni_end = eni;
+
+    hsz = 2*length; /* cons */
+
+    if (xinfo) {
+        Uint map_hsz = MAP_HEADER_FLATMAP_SZ /* 3 */;
+        if (connection_id)
+            map_size++;
+        if (node_type)
+            map_size++;
+        map_hsz += map_size; /* values */
+
+        hsz += 1 + map_size; /* common key tuple */
+        hsz += length*(3 /* 2-tuple */ + map_hsz);
+
+        if (connection_id && big_cid)
+            hsz += BIG_UINT_HEAP_SIZE*big_cid; /* big connection ids */
+    }
+
+
+    hp = HAlloc(c_p, hsz);
 #ifdef DEBUG
-    endp = hp + length*2;
+    hendp = hp + hsz;
 #endif
-    if(not_connected) {
-      for(dep = erts_not_connected_dist_entries; dep; dep = dep->next) {
-          if (dep != erts_this_dist_entry) {
-            result = CONS(hp, dep->sysname, result);
+
+    result = NIL;
+    if (!xinfo) {
+        for (eni = eni_start; eni < eni_end; eni++) {
+            result = CONS(hp, eni->name, result);
             hp += 2;
-          }
         }
-      for(dep = erts_pending_dist_entries; dep; dep = dep->next) {
-          result = CONS(hp, dep->sysname, result);
-          hp += 2;
-      }
     }
-    if(hidden)
-      for(dep = erts_hidden_dist_entries; dep; dep = dep->next) {
-	result = CONS(hp, dep->sysname, result);
-	hp += 2;
-      }
-    if(visible)
-      for(dep = erts_visible_dist_entries; dep; dep = dep->next) {
-	result = CONS(hp, dep->sysname, result);
-	hp += 2;
-      }
-    if(this) {
-	result = CONS(hp, erts_this_dist_entry->sysname, result);
-	hp += 2;
+    else {
+        Eterm keys_tuple = make_tuple(hp);
+        *(hp++) = make_arityval(map_size);
+        if (connection_id)
+            *(hp++) = am_connection_id;
+        if (node_type)
+            *(hp++) = am_node_type;
+        
+        for (eni = eni_start; eni < eni_end; eni++) {
+            flatmap_t *map;
+            Eterm cid, info_map, tuple;
+            if (connection_id) {
+                if (eni->cid ==  ~((Uint32) 0))
+                    cid = am_undefined;
+                else if (IS_USMALL(0, (Uint) eni->cid))
+                    cid = make_small((Uint) eni->cid);
+                else {
+                    cid = uint_to_big((Uint) eni->cid, hp);
+                    hp += BIG_UINT_HEAP_SIZE;
+                }
+            }
+            map = (flatmap_t *) hp;
```

Comment:
Don't like to use `erts_map_from_ks_and_vs()` since it unconditionally sorts the keys even though I already have provided them sorted, and (even worse) wont share the key tuple if multiple instances of maps with the same keys are created (which is frequent in the nodes/2 case). However, I introduced a `erts_map_from_sorted_ks_and_vs()` which does not have these issues.